Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift

```diff
@@ -0,0 +1,284 @@
+        makeDurationHistogram(name: name, unit: unit, description: description, attributes: Set(labels), buckets: buckets)
+    }
+
+    func makeValueHistogram(name: String, unit: String? = nil, description: String? = nil, labels: [(String, String)], buckets: [Double]) -> ValueHistogram {
+        makeValueHistogram(name: name, unit: unit, description: description, attributes: Set(labels), buckets: buckets)
+    }
+}
+
+struct InstrumentIdentifier: Equatable, Hashable, Sendable {
+    var name: String
+    var unit: String?
+    var description: String?
+    enum InstrumentKind { case counter, gauge, histogram }
+    var kind: InstrumentKind
+
+    private init(name: String, unit: String? = nil, description: String? = nil, kind: InstrumentKind) {
+        self.name = name.lowercased()
+        self.unit = unit
+        self.description = description
+        self.kind = kind
+    }
+
+    static func counter(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .counter)
+    }
+
+    static func gauge(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .gauge)
+    }
+
+    static func histogram(name: String, unit: String? = nil, description: String? = nil) -> Self {
+        self.init(name: name, unit: unit, description: description, kind: .histogram)
+    }
+}
+
+protocol IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { get }
+}
+
+extension Counter: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .counter(name: name, unit: unit, description: description) }
+}
+
+extension Gauge: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .gauge(name: name, unit: unit, description: description) }
+}
+
+extension Histogram: IdentifiableInstrument {
+    var instrumentIdentifier: InstrumentIdentifier { .histogram(name: name, unit: unit, description: description) }
+}
+
+protocol DuplicateRegistrationHandler: Sendable {
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>)
+}
+
+struct FatalErrorDuplicateRegistrationHandler: DuplicateRegistrationHandler {
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        fatalError("""
+        Duplicate instrument registration for name: \(newRegistration.name)
+        ---
+        Instrument \(newRegistration) conflicts with existing instruments: \(existingRegistrations).
+        """)
+    }
+}
+
+struct WarningDuplicateRegistrationHandler: DuplicateRegistrationHandler {
+    let logger: Logger
+
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        self.logger.warning("Duplicate instrument registration", metadata: [
```

Comment:
Should this be limited to only spitting out a warning once per running instance? Otherwise, it might accidentally lead to flooding the user's logging system. Maybe @ktoso can give some input on this, I know you're a fan of this warn-only-once approach.

Reply:
> It currently does it only once per duplicate instrument, which is the right thing to do, both in terms of the spec and for usability. It would be super annoying, if you had a few duplicate instruments to have to fix them one by one and re-run. That would be like your unit test suite stopping after the first failed test case.

Oh, I must have missed that part when reading the PR. One warning per instrument sounds good to me 👍

> We could consider adding a complete opt-out `.none` value for the `DuplicateRegistrationBehavior`, but I think we're pushing people in the wrong direction.

Agreed, I think we should hold off on doing it until requested by users.

> Also worth pointing out that this check and the logging is only performed after lookup has failed, which means that, if the library is used correctly (no conflicting identities) then it will take the fast path and just return the existing instrument.

Awesome 👌